File: scripts/history_prospective.py
"""Pure prospective snapshots for an already prepared history mutation.

The adapter consumes detached evidence only.  It never resolves a path, reads a
store, allocates time, or treats the candidate as committed authority.
"""
import copy
from dataclasses import replace

from . import history_adapter as A
from . import history_contract as C
from . import history_store as H
from . import history_transaction as T
from .reasoning.snapshot import Snapshot
from .pending_grounding import identity


_SUPPORTED = frozenset(('record', 'history_object', 'history_commit'))


def _fail(condition, code, detail=''):
    C._require(condition, code, detail)


def _prospective_capture(captured, mutation):
    _fail(isinstance(captured, H.Capture), 'invalid_capture')
    _fail(isinstance(mutation, T.PreparedMutation), 'invalid_mutation')
    data = mutation.to_data()
    _fail(identity(data['authority']) == identity(captured.marker), 'authority_mismatch')
    _fail(identity(data['baseline']) == identity(captured.baseline), 'baseline_mismatch')

    files = list(mutation.files)
    _fail(files, 'empty_mutation')
    by_role = {}
    for item in files:
        role = item['role']
        _fail(role in _SUPPORTED or role == 'history_evidence', 'unsupported_mutation_role', role)
        if role in ('history_object', 'history_evidence'):
            by_role.setdefault(role, []).append(item)
        else:
            _fail(role not in by_role, 'duplicate_mutation_role', role)
            by_role[role] = item
        _fail(item['before'] is None if role != 'record' else
              item['before'] == captured.entry_bytes, 'before_image_mismatch', role)
        if role not in ('record', 'history_evidence'):
            _fail(item['after'] is not None, 'invalid_mutation')
    _fail('record' in by_role and 'history_object' in by_role and 'history_commit' in by_role,
          'unsupported_mutation_shape')
    if 'history_evidence' in by_role:
        from . import history_branch
        _fail(data['receipt'].get('after', {}).get('history_branch_adoption') is not None,
              'invalid_branch_adoption_audit')
        # audit_evidences validates the retained envelope and all bindings from
        # mutation bytes alone; it performs no source Git/store reads.
        history_branch.audit_evidences(data)

    record = by_role['record']
    _fail(record['path'] == data['entry'], 'role_path_mismatch', record['path'])
    _fail(isinstance(record['after'], bytes), 'invalid_bytes')
    document = C.decode_document(record['after'])

    raw_objects = dict(captured.object_bytes)
    objects = dict(captured.objects)
    for item in by_role['history_object']:
        obj = C.validate_object(C.decode_document(item['after']))
        _fail(item['path'].endswith('.yaml'), 'role_path_mismatch', item['path'])
        key = (obj['subject'], obj['id'])
        # Adoption receipts repeat immutable objects already held by the target
        # to prove the adopted inventory. Identical bytes are not a rewrite.
        _fail(key not in raw_objects or key in captured.object_bytes
              and captured.object_bytes[key] == item['after'],
              'object_rewrite', obj['id'])
        raw_objects[key] = item['after']
        objects[obj['id']] = obj

    manifest_item = by_role['history_commit']
    manifest = C.validate_commit(C.decode_document(manifest_item['after']))
    _fail(manifest['operation'] == data['operation'], 'operation_mismatch')
    _fail(manifest['record_id'] == captured.marker['record_id'] and
          manifest['authority_generation'] == captured.marker['generation'], 'authority_mismatch')
    _fail(manifest.get('baseline_digest') == identity(captured.baseline), 'baseline_mismatch')
    _fail(manifest['parents'] == C.commit_frontier(captured.commits), 'parent_baseline_mismatch')
    commits = dict(captured.commits)
    _fail(data['operation'] not in commits, 'operation_already_prepared')
    commits[data['operation']] = manifest_item['after']
    _fail(not any(data['operation'] in generation['commits']
                  for generation in captured.inactive_generations.values()),
          'operation_collision')

    # This is the same all-or-refuse boundary used by publication.  In
    # particular, parent and object bytes are checked before any projection.
    selected = C.committed_objects(captured.marker, commits, raw_objects)
    _fail(sum(map(len, commits.values())) + sum(map(len, raw_objects.values()))
          <= H.MAX_CAPTURE_BYTES, 'history_limit')
    state = H.reduce(selected, captured.state['rules'])
    prospective_baseline = H.baseline(captured.marker, commits, state)
    _fail(C.validate_baseline(prospective_baseline) == prospective_baseline, 'invalid_baseline')
    meta = document.get('meta', {})
    _fail(isinstance(meta, dict) and identity(meta.get('history')) == identity(prospective_baseline),
          'baseline_mismatch')
    _fail(manifest.get('view_sha256') == C.sha256(record['after']), 'view_mismatch')
    virtual = replace(captured, entry_bytes=record['after'], document=document,
                   commits=commits, objects=selected, object_bytes=raw_objects,
                   state=state, baseline=prospective_baseline,
                   storage_bytes=dict(captured.storage_bytes),
                   object_paths=dict(captured.object_paths))
    # Use the canonical renderer itself so a caller cannot rehash a forged
    # record view and have it mistaken for the prepared operation's result.
    rendered = H.Store.__new__(H.Store).render(virtual, objects=selected, commits=commits)
    _fail(rendered == record['after'], 'view_mismatch')
    return virtual


def snapshot_after(captured, mutation, *, context=None, hypotheses=None, as_of=None):
    """Return an immutable supplied snapshot for ``mutation`` applied in memory."""
    for key in ('history', 'history_hypotheses', 'history_view', 'operation'):
        _fail(context is None or key not in context, 'duplicate_prospective_context', key)
    virtual = _prospective_capture(captured, mutation)
    adapted = A.from_store_capture(virtual)
    supplied = copy.deepcopy(context) if context is not None else {}
    supplied['read_mode'] = 'supplied'
    supplied['operation'] = {
        'version': 1,
        'phase': 'prospective',
        'kind': 'prepared_history',
        'original_baseline': copy.deepcopy(captured.baseline),
        'mutation_operation': mutation.to_data()['operation'],
    }
    return adapted.snapshot(context=supplied, hypotheses=hypotheses, as_of=as_of)


def assess(captured, mutation, *, hypotheses=None, as_of=None):
    """Compare committed authority with its prepared candidate, never a live overlay.

    Physical hypotheses may be supplied from the caller's verified capture. Named
    history hypotheses are reconstructed independently on each side. Pending
    contributions and publication policy remain the caller's separate boundary.
    """
    from .reasoning.context import CapturedAssessment
    from .reasoning.operations import findings
    context = {'read_mode': 'supplied', 'operation_scope': 'committed_history'}
    before = A.from_store_capture(captured).snapshot(context=context, hypotheses=hypotheses, as_of=as_of)
    after = snapshot_after(captured, mutation, context=context, hypotheses=hypotheses, as_of=before.to_data()['as_of'])
    before_context = CapturedAssessment.from_snapshot(before)
    after_context = CapturedAssessment.from_snapshot(after)
    old, new = findings(before_context), findings(after_context)
    introduced = {key: sorted(set(new[key]) - set(old[key]))
                  for key in ('falsified', 'holes', 'moved', 'notes')}
    return {'before': before_context, 'after': after_context,
            'before_findings': old, 'after_findings': new, 'introduced': introduced}

File: native/src/history_prospective.rs
//! Pure proposed history worlds; detached mutations never become accepted authority.
use crate::{
    Result, history_adapter as D,
    history_authoring::{empty, n, obj, s, strings},
    history_authority as A,
    history_capture::{self as H, Capture},
    history_contract::*,
    history_reduce,
    history_transaction::PreparedMutation,
    history_view::{self as View, list, map_mut},
    history_yaml as Y,
    identity::sha256,
    reasoning_context::CapturedAssessment,
    reasoning_operations::findings,
    reasoning_runtime::{OperationalBounds, Runtime},
    reasoning_snapshot::{CaptureOptions, Snapshot},
    require,
    value::TypedValue as V,
};
use std::collections::BTreeSet;

pub fn capture_after(captured: &Capture, mutation: &PreparedMutation) -> Result<Capture> {
    let data = mutation.to_data();
    let data = map(&data)?;
    require(
        data["authority"].digest()? == captured.marker.digest()?,
        "authority_mismatch",
    )?;
    require(
        data["baseline"].digest()? == captured.baseline.digest()?,
        "baseline_mismatch",
    )?;
    let mut record = None;
    let mut commit = None;
    let mut objects = vec![];
    for file in mutation.files() {
        match file.role.as_str() {
            "record" => {
                require(record.replace(file).is_none(), "duplicate_mutation_role")?;
            }
            "history_commit" => {
                require(commit.replace(file).is_none(), "duplicate_mutation_role")?;
            }
            "history_object" => objects.push(file),
            // Branch evidence requires the separate retained adoption audit.
            // Refuse until that audit is available; a hash is not its substitute.
            "history_evidence" => return Err(error("branch_adoption_audit_unsupported")),
            _ => return Err(error("unsupported_mutation_role")),
        }
        require(
            if file.role == "record" {
                file.before.as_ref() == Some(&captured.entry_bytes)
            } else {
                file.before.is_none()
            },
            "before_image_mismatch",
        )?;
        require(file.after.is_some(), "invalid_mutation")?;
    }
    require(!objects.is_empty(), "unsupported_mutation_shape")?;
    let record = record.ok_or_else(|| error("unsupported_mutation_shape"))?;
    let commit = commit.ok_or_else(|| error("unsupported_mutation_shape"))?;
    require(record.path == text(&data["entry"])?, "role_path_mismatch")?;
    let mut virtual_capture = captured.clone();
    virtual_capture.entry_bytes = record.after.clone().unwrap();
    virtual_capture.document = Y::decode_document(&virtual_capture.entry_bytes)?;
    for file in objects {
        let raw = file.after.as_ref().unwrap();
        let object = Y::decode_document(raw)?;
        validate_object(&object)?;
        let object = map(&object)?;
        let key = (
            text(&object["subject"])?.to_owned(),
            text(&object["id"])?.to_owned(),
        );
        require(
            virtual_capture
                .object_bytes
                .get(&key)
                .is_none_or(|old| old == raw),
            "object_rewrite",
        )?;
        virtual_capture.object_bytes.insert(key, raw.clone());
    }
    let manifest = Y::decode_document(commit.after.as_ref().unwrap())?;
    A::validate_commit(&manifest)?;
    let manifest = map(&manifest)?;
    require(
        manifest["operation"] == data["operation"],
        "operation_mismatch",
    )?;
    require(
        manifest["record_id"] == map(&captured.marker)?["record_id"]
            && manifest["authority_generation"] == map(&captured.marker)?["generation"],
        "authority_mismatch",
    )?;
    require(
        field(manifest, "baseline_digest")? == &s(&captured.baseline.digest()?),
        "baseline_mismatch",
    )?;
    require(
        manifest["parents"] == V::Map(A::commit_frontier(&captured.commits)?),
        "parent_baseline_mismatch",
    )?;
    let operation = text(&data["operation"])?;
    require(
        !virtual_capture.commits.contains_key(operation),
        "operation_already_prepared",
    )?;
    for generation in captured.inactive_generations.values() {
        require(
            !generation.commits.contains_key(operation),
            "operation_collision",
        )?;
    }
    virtual_capture
        .commits
        .insert(operation.into(), commit.after.clone().unwrap());
    virtual_capture.objects = A::committed_objects(
        &captured.marker,
        &virtual_capture.commits,
        &virtual_capture.object_bytes,
    )?;
    let total: usize = virtual_capture
        .commits
        .values()
        .map(Vec::len)
        .sum::<usize>()
        .saturating_add(virtual_capture.object_bytes.values().map(Vec::len).sum());
    require(total <= 64 * 1024 * 1024, "history_limit")?;
    let selected = virtual_capture
        .object_bytes
        .iter()
        .filter(|((_, id), _)| virtual_capture.objects.contains_key(id))
        .map(|(k, v)| (k.clone(), v.clone()))
        .collect();
    virtual_capture.state =
        history_reduce::reduce_bytes(&selected, Some(map(&map(&captured.state)?["rules"])?), None)?;
    virtual_capture.baseline = H::baseline(
        &captured.marker,
        &virtual_capture.commits,
        &virtual_capture.state,
    )?;
    require(
        field(
            map(field(map(&virtual_capture.document)?, "meta")?)?,
            "history",
        )?
        .digest()?
            == virtual_capture.baseline.digest()?,
        "baseline_mismatch",
    )?;
    require(
        field(manifest, "view_sha256")? == &s(&sha256(&virtual_capture.entry_bytes)),
        "view_mismatch",
    )?;
    require(
        View::render(
            &virtual_capture,
            &virtual_capture.objects,
            &virtual_capture.object_bytes,
            &virtual_capture.commits,
        )? == virtual_capture.entry_bytes,
        "view_mismatch",
    )?;
    Ok(virtual_capture)
}

pub fn snapshot_after(
    captured: &Capture,
    mutation: &PreparedMutation,
    mut options: CaptureOptions,
) -> Result<Snapshot> {
    let mut context = options
        .context
        .take()
        .filter(|v| *v != V::Null)
        .unwrap_or_else(empty);
    require(
        !["history", "history_hypotheses", "history_view", "operation"]
            .iter()
            .any(|k| map(&context).is_ok_and(|m| m.contains_key(*k))),
        "duplicate_prospective_context",
    )?;
    let virtual_capture = capture_after(captured, mutation)?;
    map_mut(&mut context)?.insert("read_mode".into(), s("supplied"));
    map_mut(&mut context)?.insert(
        "operation".into(),
        obj([
            ("version", n("1")),
            ("phase", s("prospective")),
            ("kind", s("prepared_history")),
            ("original_baseline", captured.baseline.clone()),
            (
                "mutation_operation",
                map(&mutation.to_data())?["operation"].clone(),
            ),
        ]),
    );
    options.context = Some(context);
    D::from_store_capture(&virtual_capture)?.snapshot(options)
}

pub struct Assessment {
    pub before: CapturedAssessment,
    pub after: CapturedAssessment,
    pub before_findings: V,
    pub after_findings: V,
    pub introduced: V,
}
pub fn assess(
    captured: &Capture,
    mutation: &PreparedMutation,
    hypotheses: Option<&V>,
    as_of: Option<&V>,
    runtime: Option<&Runtime>,
) -> Result<Assessment> {
    let context = obj([
        ("read_mode", s("supplied")),
        ("operation_scope", s("committed_history")),
    ]);
    let before = D::from_store_capture(captured)?.snapshot(CaptureOptions {
        context: Some(context.clone()),
        hypotheses: hypotheses.cloned(),
        as_of: as_of.cloned(),
        ..Default::default()
    })?;
    let after = snapshot_after(
        captured,
        mutation,
        CaptureOptions {
            context: Some(context),
            hypotheses: hypotheses.cloned(),
            as_of: Some(map(before.data())?["as_of"].clone()),
            ..Default::default()
        },
    )?;
    let before = CapturedAssessment::from_snapshot(
        before,
        None,
        "focused-review/v1",
        runtime,
        OperationalBounds::default(),
        None,
    )?;
    let after = CapturedAssessment::from_snapshot(
        after,
        None,
        "focused-review/v1",
        runtime,
        OperationalBounds::default(),
        None,
    )?;
    let before_findings = findings(&before)?;
    let after_findings = findings(&after)?;
    let mut introduced = Map::new();
    for key in ["falsified", "holes", "moved", "notes"] {
        let old = list(&map(&before_findings)?[key])?
            .iter()
            .map(|v| text(v).map(str::to_owned))
            .collect::<Result<BTreeSet<_>>>()?;
        let new = list(&map(&after_findings)?[key])?
            .iter()
            .map(|v| text(v).map(str::to_owned))
            .collect::<Result<BTreeSet<_>>>()?;
        introduced.insert(key.into(), strings(new.difference(&old).cloned()));
    }
    Ok(Assessment {
        before,
        after,
        before_findings,
        after_findings,
        introduced: V::Map(introduced),
    })
}
